[DOC] Update TraceQL query editor doc (#91910)
grafana/grafana: docs/sources/shared/datasources/tempo-editor-traceql.md

To learn more about how to query by TraceQL, refer to the [TraceQL documentation](https://grafana.com/docs/tempo/latest/traceql/).

> The TraceQL query editor **Explore** in Grafana lets you search by trace ID and write TraceQL queries using autocomplete.

Reviewer: I think you mean _in_ Explore? If so:
```suggestion
The TraceQL query editor in Grafana **Explore** lets you search by trace ID and write TraceQL queries using autocomplete.
```

![The TraceQL query editor](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-traceql-v11.png)

## Before you begin

This feature is automatically available in Grafana 10 (and newer) and Grafana Cloud.

To use the TraceQL query editor in self-hosted Grafana 9.3.2 and older, you need to [enable the `traceqlEditor` feature toggle](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/).

### Streaming and gRPC

> If trying to query a self-managed Grafana Tempo or Grafana Enterprise Traces database with a gateway, such as nginx, in front of it from your hosted Grafana, that gateway (for example, nginx) must allow gRPC connections.

Reviewer: ```suggestion
If you're trying to query a self-managed Grafana Tempo or Grafana Enterprise Traces database with a gateway, such as nginx, in front of it from your hosted Grafana, that gateway (for example, nginx) must allow gRPC connections.
```

If it doesn't, streaming won't work and queries will fail to return results.

If you can't configure your gateway to allow gRPC, deactivate streaming in your hosted Grafana.

> In Grafana 11.2 and newer, you can deactivate **Streaming** in your Tempo data source settings from **Connections** > **Data sources** in the Grafana menu.

Reviewer: ```suggestion
In Grafana 11.2 and newer, you can disable the **Streaming** option in your Tempo data source settings from **Connections** > **Data sources** in the Grafana menu.
```

Author: Aren't we moving away from wording like "enable" and "disable"?

Reviewer: My bad. Conflating style guides in my head. 
```suggestion
In Grafana 11.2 and newer, you can deactivate the **Streaming** option in your Tempo data source settings from **Connections** > **Data sources** in the Grafana menu.
```

Author: I appreciate all of your comments. Thank you for the thorough review! You've found a bunch of stuff. :)

> In Grafana 11.2 and newer, you can deactivate the **Streaming** option in your Tempo data source settings from **Connections** > **Data sources** in the Grafana menu.

Reviewer: ```suggestion
In Grafana 11.2 and newer, you can deactivate the **Streaming** option in your Tempo data source settings from **Connections** > **Data sources** in the Grafana main menu.
```

To access the query editor, follow these steps:

1. Sign into Grafana or Grafana Cloud.

> 1. Select **Explore** from the menu.

Reviewer: ```suggestion
1. Select **Explore** from the main menu.
```

1. Select a Tempo data source.
1. Select the **TraceQL** tab.
1. Start your query on the text line by entering `{`. For help with TraceQL syntax, refer to the [Construct a TraceQL query documentation](https://grafana.com/docs/tempo/latest/traceql/#construct-a-traceql-query).

>    Optional: Select **Copy query from Search** to transfer a builder query to the editor.

Reviewer: ```suggestion

   Optional: Select **Copy query from Search** to transfer a builder query to the editor.
   
```

1. Optional: Use the **Time picker** drop-down list to change the time and range for the query (refer to the [documentation for instructions](https://grafana.com/docs/grafana/latest/dashboards/use-dashboards/#set-dashboard-time-range)).

> 1. Once you have finished your query, select **Run query**.

Reviewer: ```suggestion
1. Once you've finished your query, select **Run query**.
```

> This video provides and example of creating a TraceQL query using the custom tag grouping.

Reviewer: ```suggestion
This video provides an example of creating a TraceQL query using the custom tag grouping.
```

1. Enter the trace ID into the query field. For example: `41928b92edf1cdbe0ba6594baee5ae9`
1. Click **Run query** or use the keyboard shortcut Shift + Enter.

![Search for a trace ID using the TraceQL query editor](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-trace-id-v11.png)

## Use autocomplete to write queries

You can use the query editor’s autocomplete suggestions to write queries.

> The editor detects span sets to provide relevant autocomplete options.

Reviewer: ```suggestion
The editor detects spansets to provide relevant autocomplete options.
```

> It uses regular expressions (regex) to detect where it's inside a spanset and provide attribute names, scopes, intrinsic names, logic operators, or attribute values from the Tempo API, depending on what's expected for the current situation.

Reviewer: ```suggestion
It uses regular expressions (regex) to detect where it is inside a spanset and provide attribute names, scopes, intrinsic names, logic operators, or attribute values from the Tempo API, depending on what's expected for the current situation.
```

## View query results

Query results for both the editor and the builder return in a table.
You can interact with the data listed to view more details.

Selecting the trace ID from the returned results opens a trace diagram.
Selecting a span from the returned results opens a trace diagram and reveals the relevant span in the trace diagram.

> 

Reviewer: I know it's a lot of work, but the image is so busy that it's hard to parse the things described in lines 100-104. I'd suggest putting the image on line 106 here and then formatting lines 100-104 as a bullet list, each with an accompanying much smaller detail image to show the specific thing you're talking about. Hopefully you'd just be able to take cuts from the larger image to do this.

Author: I've actually found some text that has a better explanation in the TraceQL doc in Tempo OSS. I'm goign to use that and the screenshot from it. I think it will be a lot clearer.

Author: I rewrote this section to point to the doc that explains span details instead of duplicating it here. I also added some links that point to more details for other concepts.

In the trace diagram, the bold text on the left side of each span indicates the service name, for example `otel-demo: HTTP POST`.
It's hidden when subsequent spans have the same service name (nested spans).
Each service has a color assigned to it, which is visible as a highlight and underline and timeline in the graph.
Spans with the same color belong to the same service.
The grey text to the right of the service name indicates the span name.

![Query editor showing span results](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-ed-example-v11-a.png)

### Focus on traces or spans

Under **Options**, you can choose to display the table as **Traces** or **Spans** focused.

> When the Table Type option is set to Spans, the traces and span sets are flattened into a list of spans.

Reviewer: ```suggestion
When the **Table Type** option is set to **Spans**, the traces and span sets are flattened into a list of spans.
```

> When the **Table Type** option is set to **Spans**, the traces and span sets are flattened into a list of spans.

Reviewer: ```suggestion
When the **Table Type** option is set to **Spans**, the traces and spansets are flattened into a list of spans.
```
